cross compiling failed

Hi there,
I tried to cross-compile rummqtt and got the following output.

```
roker/target/armv7-unknown-linux-gnueabihf/release/deps/libthiserror-eef2e33c746f6e56.rmeta --cap-lints allow`
error[E0308]: mismatched types
   --> /home/xx/.cargo/registry/src/github.com-1ecc6299db9ec823/pprof-0.4.1/src/profiler.rs:143:47
    |
143 |             write_thread_name(current_thread, &mut name);
    |                                               ^^^^^^^^^ expected slice `[u8]`, found array `[i8; 16]`
    |
    = note: expected mutable reference `&mut [u8]`
               found mutable reference `&mut [i8; 16]`

error[E0606]: casting `&mut [i8; 16]` as `*mut [u8]` is invalid
   --> /home/fwa/.cargo/registry/src/github.com-1ecc6299db9ec823/pprof-0.4.1/src/profiler.rs:141:28
    |
141 |             let name_ptr = &mut name as *mut [libc::c_char] as *mut libc::c_char;
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0308, E0606.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `pprof`
```


Can you verify this?
Thanks in advance.

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -6,10 +6,10 @@ jobs:
   lint:
     name: Format and Clippy
     strategy:
-       matrix:
-         os: [ubuntu-latest, macos-latest]
-         toolchain: [stable, nightly]
-         target: [x86_64-unknown-linux-gnu, aarch64-unknown-linux-gnu, x86_64-apple-darwin, aarch64-apple-darwin]
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+        toolchain: [stable, nightly]
+        target: [x86_64-unknown-linux-gnu, aarch64-unknown-linux-gnu, x86_64-apple-darwin, aarch64-apple-darwin]
     runs-on: ${{ matrix.os }}
 
     steps:
@@ -41,7 +41,9 @@ jobs:
     name: Test
     strategy:
       matrix:
-        os: [ubuntu-latest]
+        os: [ubuntu-latest, macos-latest]
+        toolchain: [stable, nightly]
+        target: [x86_64-unknown-linux-gnu, aarch64-unknown-linux-gnu, x86_64-apple-darwin, aarch64-apple-darwin]
     runs-on: ${{ matrix.os }}
 
     steps:

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -35,9 +35,10 @@ jobs:
     name: Build
     strategy:
       matrix:
+        os: [ubuntu-latest, macos-latest]
         toolchain: [stable, nightly]
         target: [x86_64-unknown-linux-gnu, aarch64-unknown-linux-gnu, x86_64-apple-darwin, aarch64-apple-darwin]
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
 
     steps:
       - name: Checkout sources

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -8,6 +8,8 @@ jobs:
     strategy:
        matrix:
          os: [ubuntu-latest, macos-latest]
+         toolchain: [stable, nightly]
+         target: [x86_64-unknown-linux-gnu, aarch64-unknown-linux-gnu, x86_64-apple-darwin, aarch64-apple-darwin]
     runs-on: ${{ matrix.os }}
 
     steps:
@@ -18,7 +20,8 @@ jobs:
         uses: actions-rs/toolchain@v1
         with:
           profile: minimal
-          toolchain: stable
+          toolchain: ${{ matrix.toolchain }}
+          target: ${{ matrix.target }}
           override: true
           components: rustfmt, clippy
 
@@ -49,7 +52,8 @@ jobs:
         uses: actions-rs/toolchain@v1
         with:
           profile: minimal
-          toolchain: stable
+          toolchain: ${{ matrix.toolchain }}
+          target: ${{ matrix.target }}
           override: true
 
       - name: Run cargo test

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -5,12 +5,7 @@ name: build
 jobs:
   lint:
     name: Format and Clippy
-    strategy:
-      matrix:
-        os: [ubuntu-latest, macos-latest]
-        toolchain: [stable, nightly]
-        target: [x86_64-unknown-linux-gnu, aarch64-unknown-linux-gnu, x86_64-apple-darwin, aarch64-apple-darwin]
-    runs-on: ${{ matrix.os }}
+    runs-on: ubuntu-latest
 
     steps:
       - name: Checkout sources
@@ -20,8 +15,7 @@ jobs:
         uses: actions-rs/toolchain@v1
         with:
           profile: minimal
-          toolchain: ${{ matrix.toolchain }}
-          target: ${{ matrix.target }}
+          toolchain: stable
           override: true
           components: rustfmt, clippy
 

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -31,14 +31,13 @@ jobs:
           command: clippy
           args: --all-features -- -D warnings
 
-  test:
-    name: Test
+  build:
+    name: Build
     strategy:
       matrix:
-        os: [ubuntu-latest, macos-latest]
         toolchain: [stable, nightly]
         target: [x86_64-unknown-linux-gnu, aarch64-unknown-linux-gnu, x86_64-apple-darwin, aarch64-apple-darwin]
-    runs-on: ${{ matrix.os }}
+    runs-on: ubuntu-latest
 
     steps:
       - name: Checkout sources
@@ -50,6 +49,31 @@ jobs:
           profile: minimal
           toolchain: ${{ matrix.toolchain }}
           target: ${{ matrix.target }}
+          override: true
+
+      - name: Run cargo build
+        uses: actions-rs/cargo@v1
+        with:
+          command: build
+          args: --all-features --target ${{ matrix.target }}
+
+  test:
+    name: Test
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+        toolchain: [stable, nightly]
+    runs-on: ${{ matrix.os }}
+
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v2
+
+      - name: Install Rust toolchain
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: ${{ matrix.toolchain }}
           override: true
 
       - name: Run cargo test

--- a/examples/malloc_hook.rs
+++ b/examples/malloc_hook.rs
@@ -6,6 +6,7 @@ use pprof;
 use std::ffi::c_void;
 
 #[cfg(not(target_os = "linux"))]
+#[allow(clippy::wrong_self_convention)]
 static mut __malloc_hook: Option<extern "C" fn(size: usize) -> *mut c_void> = None;
 
 extern "C" {

--- a/examples/malloc_hook.rs
+++ b/examples/malloc_hook.rs
@@ -5,7 +5,11 @@ extern crate libc;
 use pprof;
 use std::ffi::c_void;
 
+#[cfg(not(target_os = "linux"))]
+static mut __malloc_hook: Option<extern "C" fn(size: usize) -> *mut c_void> = None;
+
 extern "C" {
+    #[cfg(target_os = "linux")]
     static mut __malloc_hook: Option<extern "C" fn(size: usize) -> *mut c_void>;
 
     fn malloc(size: usize) -> *mut c_void;

--- a/src/collector.rs
+++ b/src/collector.rs
@@ -343,6 +343,7 @@ mod tests {
     }
 
     #[cfg(not(target_os = "linux"))]
+    #[allow(clippy::wrong_self_convention)]
     static mut __malloc_hook: Option<extern "C" fn(size: usize) -> *mut c_void> = None;
 
     extern "C" {

--- a/src/collector.rs
+++ b/src/collector.rs
@@ -342,7 +342,11 @@ mod tests {
         }
     }
 
+    #[cfg(not(target_os = "linux"))]
+    static mut __malloc_hook: Option<extern "C" fn(size: usize) -> *mut c_void> = None;
+
     extern "C" {
+        #[cfg(target_os = "linux")]
         static mut __malloc_hook: Option<extern "C" fn(size: usize) -> *mut c_void>;
 
         fn malloc(size: usize) -> *mut c_void;

--- a/src/profiler.rs
+++ b/src/profiler.rs
@@ -214,6 +214,7 @@ impl Profiler {
     }
 }
 
+#[cfg(target_os = "linux")]
 #[cfg(test)]
 mod tests {
     use super::*;

--- a/src/profiler.rs
+++ b/src/profiler.rs
@@ -137,7 +137,7 @@ extern "C" fn perf_signal_handler(_signal: c_int) {
             }
 
             let current_thread = unsafe { libc::pthread_self() };
-            let mut name = [0 as libc::c_char; MAX_THREAD_NAME];
+            let mut name = [0; MAX_THREAD_NAME];
             let name_ptr = &mut name as *mut [libc::c_char] as *mut libc::c_char;
 
             write_thread_name(current_thread, &mut name);

--- a/src/profiler.rs
+++ b/src/profiler.rs
@@ -221,6 +221,7 @@ mod tests {
     use std::ffi::c_void;
 
     #[cfg(not(target_os = "linux"))]
+    #[allow(clippy::wrong_self_convention)]
     static mut __malloc_hook: Option<extern "C" fn(size: usize) -> *mut c_void> = None;
 
     extern "C" {

--- a/src/profiler.rs
+++ b/src/profiler.rs
@@ -214,14 +214,17 @@ impl Profiler {
     }
 }
 
-#[cfg(target_os = "linux")]
 #[cfg(test)]
 mod tests {
     use super::*;
     use std::cell::RefCell;
     use std::ffi::c_void;
 
+    #[cfg(not(target_os = "linux"))]
+    static mut __malloc_hook: Option<extern "C" fn(size: usize) -> *mut c_void> = None;
+
     extern "C" {
+        #[cfg(target_os = "linux")]
         static mut __malloc_hook: Option<extern "C" fn(size: usize) -> *mut c_void>;
 
         fn malloc(size: usize) -> *mut c_void;

--- a/src/profiler.rs
+++ b/src/profiler.rs
@@ -137,7 +137,7 @@ extern "C" fn perf_signal_handler(_signal: c_int) {
             }
 
             let current_thread = unsafe { libc::pthread_self() };
-            let mut name = [0_i8; MAX_THREAD_NAME];
+            let mut name = [0 as libc::c_char; MAX_THREAD_NAME];
             let name_ptr = &mut name as *mut [libc::c_char] as *mut libc::c_char;
 
             write_thread_name(current_thread, &mut name);

